Support for flattened NPM 3-style directories

GoGoCarl opened the pull request:
Check the NPM version before deciding the source and target paths

GoGoCarl committed:
Support for flattened NPM 3-style directories

Check the NPM version before deciding the source and target paths

GoGoCarl commented:
If some or all of this is useful to you in reference to #4, feel free.  I have done very limited testing, but I am currently using this to workaround the issue myself.

wearhere commented on src/index.js:
```diff
@@ -31,15 +33,27 @@ function installFiles(sourceDir, done) {
+  }
+
+  if (fileInstallingPackagePath.match(".+" + process.env.npm_package_name + "$")) {
+    console.log("[install-files]: Target = self, skipping install")
```
oh this is a good idea

wearhere commented on src/index.js:
```diff
@@ -31,15 +33,27 @@ function installFiles(sourceDir, done) {
+
+  if (fileInstallingPackagePath.match(".+" + process.env.npm_package_name + "$")) {
+    console.log("[install-files]: Target = self, skipping install")
+    return;
```
please still call `process.nextTick(done);`

GoGoCarl replied:
Ah, good call.

wearhere commented on src/index.js:
```diff
@@ -31,15 +33,27 @@ function installFiles(sourceDir, done) {
-  if (!destinationDir) {
+  var source, target;
+  switch (npmv.majorVersion()) {
+    case '3':
```
Let's actually invert the cases (`default -> '2'`, `'3' -> default`) since presumably npm will keep their format v3 format going forward.

It'd probably be good to make the new `'2'` branch fall through to a `'1'` branch too although I actually don't know what the v1 format was or even if anyone uses v1 anymore haha.

wearhere replied:
(Hopefully the major version can still be determined in v2?)

GoGoCarl replied:
I was thinking about that too and came to the same conclusion -- I wasn't certain (and didn't look) to see if the version could be detected in v2.  Since I *knew* I could get a version from v3, that was the one I checked explicitly.

Similarly, no clue about v1 -- I think if you're still on that it might be time to upgrade before installing any other modules!

wearhere replied:
I just tested and `process.env.npm_config_user_agent` is set/of the same format in v2, so all clear to invert the cases.

wearhere commented on src/npm-version.js:
```diff
@@ -0,0 +1,29 @@
+  var version = undefined;
+  if (process.env.npm_config_user_agent && process.env.npm_config_user_agent.match(/.*npm\/.+/)) {
+    var agent = process.env.npm_config_user_agent.split(' ')
+    for (var index in agent) {
```
Please do `var… of` rather than `in`, see [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of#Difference_between_for...of_and_for...in).

GoGoCarl replied:
Learn something new every day!

wearhere commented:
Oh awesome, thanks @GoGoCarl! Definitely looks like an improvement on the old behavior.

wearhere reviewed (changes requested):
just a few comments, thanks!

GoGoCarl committed:
Explicitly handle older versions of NPM

Default to NPMv3-style paths, and expliciltly handle older
versions.  Also, some code cleanup.

GoGoCarl committed:
Check for null, not undefined npm major version

GoGoCarl commented:
@wearhere Updated!

wearhere commented:
Great, thanks again @GoGoCarl! Published as v1.1.0.
